packages/insomnia-app/app/ui/context/nunjucks/use-render-functions.ts

```diff
@@ -0,0 +1,63 @@
+import { useCallback, useMemo } from 'react';
+import { useSelector } from 'react-redux';
+
+import { getRenderContext, getRenderContextAncestors, HandleGetRenderContext, HandleRender, render } from '../../../common/render';
+import { NUNJUCKS_TEMPLATE_GLOBAL_PROPERTY_NAME } from '../../../templating';
+import { getKeys } from '../../../templating/utils';
+import { selectActiveEnvironment, selectActiveRequest, selectActiveWorkspace } from '../../redux/selectors';
+
+/**
+ * @deprecated this hook is intended to only be used by `NunjucksRenderFunctionProvider`
+ *
+ * Use `useNunjucksRenderFunctions` or `useGatedNunjucksRenderFunctions` instead
+ */
+export const useRenderFunctions = () => {
+  const getRenderContextPromiseCache = useMemo(() => ({}), []);
+
+  const environmentId = useSelector(selectActiveEnvironment)?._id;
+  const request = useSelector(selectActiveRequest);
+  const workspace = useSelector(selectActiveWorkspace);
+
+  const fetchRenderContext = useCallback(async () => {
+    const ancestors = await getRenderContextAncestors(request || workspace);
+    return getRenderContext({
+      request: request || undefined,
+      environmentId,
+      ancestors,
+    });
+  }, [environmentId, request, workspace]);
+
+  const handleGetRenderContext: HandleGetRenderContext = useCallback(async () => {
+    const context = await fetchRenderContext();
+    const keys = getKeys(context, NUNJUCKS_TEMPLATE_GLOBAL_PROPERTY_NAME);
+    return { context, keys };
+  }, [fetchRenderContext]);
+
+  /**
+   * Heavily optimized render function
+   *
+   * @param text - template to render
+   * @param contextCacheKey - if rendering multiple times in parallel, set this
+   * @returns {Promise}
+   * @private
+   */
+  const handleRender: HandleRender = useCallback(async <T>(obj: T, contextCacheKey: string | null = null) => {
```

Comment:
We could use undefined instead of null here
```suggestion
  const handleRender: HandleRender = useCallback(async <T>(obj: T, contextCacheKey?: string) => {
```
 or even better not accept an undefined cache key in the function

Reply:
Same as https://github.com/Kong/insomnia/pull/4248#discussion_r760670168, just refactored as-is from app.tsx